server/src/http/broadcaster.rs

```diff
@@ -0,0 +1,207 @@
+            .and_modify(|group| {
+                group.clients.push(tx.clone());
+            })
+            .or_insert(ClientGroup {
+                clients: vec![tx.clone()],
+                filter_set,
+                token,
+            });
+
+        Ok(Sse::from_infallible_receiver(rx))
+    }
+
+    fn get_query_filters(
+        filter_query: Query<FeatureFilters>,
+        token: &EdgeToken,
+    ) -> FeatureFilterSet {
+        let query_filters = filter_query.into_inner();
+
+        let filter_set = if let Some(name_prefix) = query_filters.name_prefix {
+            FeatureFilterSet::from(Box::new(name_prefix_filter(name_prefix)))
+        } else {
+            FeatureFilterSet::default()
+        }
+        .with_filter(project_filter(token));
+        filter_set
+    }
+
+    async fn resolve_features(
+        &self,
+        validated_token: &EdgeToken,
+        filter_set: Query<FeatureFilters>,
+        query: FlagQuery,
+    ) -> EdgeJsonResult<ClientFeatures> {
+        let filter_set = Broadcaster::get_query_filters(filter_set.clone(), validated_token);
+
+        let features = self
+            .features_cache
+            .get(&cache_key(validated_token))
+            .map(|client_features| filter_client_features(&client_features, &filter_set));
+
+        match features {
+            Some(features) => Ok(Json(ClientFeatures {
+                query: Some(query),
+                ..features
+            })),
+            // Note: this is a simplification for now, using the following assumptions:
+            // 1. We'll only allow streaming in strict mode
+            // 2. We'll check whether the token is subsumed *before* trying to add it to the broadcaster
+            // If both of these are true, then we should never hit this case (if Thomas's understanding is correct).
+            None => todo!(),
+        }
+    }
+
+    /// Broadcast new features to all clients.
+    pub async fn broadcast(&self) {
+        let mut client_events = Vec::new();
+        for entry in self.active_connections.iter() {
+            let (_query, group) = entry.pair();
+            let filter_set = Broadcaster::get_query_filters(group.filter_set.clone(), &group.token);
+            let features = self
+                .features_cache
+                .get(&cache_key(&group.token))
+                .map(|client_features| filter_client_features(&client_features, &filter_set));
+            let event: Event = sse::Data::new_json(&features)
+                .unwrap()
+                .event("unleash-updated")
+                .into();
+
+            for client in &group.clients {
+                client_events.push((client.clone(), event.clone()));
+            }
+        }
+        // try to send to all clients, ignoring failures
+        // disconnected clients will get swept up by `remove_stale_clients`
+        let send_events = client_events
+            .iter()
+            .map(|(client, event)| client.send(event.clone()));
```

Comment:
Do we need this clone?

Reply:
At the moment, yes we do. Because `iter` returns a pair of references we only get an `&Event`, and the send function needs to own the event. So I don't think it's possible to send these events without cloning the way we do it now.

As a potential refactor (which would probably be beneficial), we could:

- Loop through the groups and collect the right features for each.
- Loop through the clients in each group and generate the event for each client (because you can generate the event from a reference to the features).

There might be smarter ways too, but let's look at that as its own task 😄